### ISSUE TITLE
Add impatient back

### ISSUE THREAD

**Issue by nullchilly** [NvChad/NvChad#1427]
**Is your feature request related to a problem? Please describe.**
Pretty pointless to remove it in the first place to be honest
15ms without vs 21ms with impatient for me
Although it doesn't improve startuptime it does optimize real-time loading by a big margin

**Describe the solution you'd like**
In any case adding this doesn't affect `--startuptime` or `hyperfine` if you want to trick the profiler
```lua
  ["lewis6991/impatient.nvim"] = {
    event = "VimEnter",
    config = function()
      vim.defer_fn(function()
        local present, impatient = pcall(require, "impatient")
        if present then impatient.enable_profile() end
      end, 0)
    end,
  },
```
impatient will cache the lazy-loaded plugins:
![image](https://user-images.githubusercontent.com/56817415/183072770-de5bdf21-c97c-4757-ac6d-1955614c4d24.png)

**Comment by siduck** [comment]
if it increases startuptime then I cannot add it ._.

**Comment by nullchilly** [comment]
It is the same as lazyloading

```sh
> hyperfine "nvim +q"
Benchmark 1: nvim +q
  Time (mean ± σ):      25.6 ms ±   0.8 ms    [User: 10.6 ms, System: 5.6 ms]
  Range (min … max):    24.3 ms …  29.5 ms    110 runs
```
```
times in msec
 clock   self+sourced   self:  sourced script
 clock   elapsed:              other lines

000.005  000.005: --- NVIM STARTING ---
000.066  000.061: event init
000.126  000.060: early init
000.174  000.048: locale set
000.215  000.041: init first window
000.462  000.246: inits 1
000.474  000.013: window checked
000.477  000.002: parsing arguments
000.762  000.049  000.049: require('vim.shared')
000.842  000.035  000.035: require('vim._meta')
000.844  000.079  000.045: require('vim._editor')
000.844  000.153  000.025: require('vim._init_packages')
000.846  000.216: init lua interpreter
000.872  000.026: expanding arguments
000.898  000.025: inits 2
001.046  000.149: init highlight
001.047  000.001: waiting for UI
001.738  000.691: done waiting for UI
001.747  000.009: init screen for UI
001.767  000.020: init default mappings
001.782  000.015: init default autocommands
002.138  000.067  000.067: sourcing /usr/local/share/nvim/runtime/ftplugin.vim
002.196  000.028  000.028: sourcing /usr/local/share/nvim/runtime/indent.vim
002.234  000.008  000.008: sourcing /usr/share/nvim/archlinux.vim
002.236  000.021  000.014: sourcing /etc/xdg/nvim/sysinit.vim
002.632  000.357  000.357: require('core')
002.829  000.136  000.136: require('core.utils')
003.061  000.209  000.209: require('core.mappings')
003.062  000.232  000.023: require('core.default_config')
004.206  001.572  001.204: require('core.options')
004.393  000.167  000.167: require('core.packer')
004.545  000.004  000.004: require('vim.F')
004.947  000.266  000.266: require('core.lazy_load')
005.560  000.106  000.106: require('packer.util')
005.584  000.632  000.527: require('packer')
006.289  000.203  000.203: require('packer.log')
006.293  000.301  000.098: require('packer.async')
006.507  000.073  000.073: require('packer.result')
006.510  000.216  000.143: require('packer.jobs')
006.518  000.714  000.197: require('packer.plugin_utils')
006.693  000.168  000.168: require('packer.snapshot')
006.834  002.421  000.638: require('plugins')
006.888  004.641  000.124: sourcing /tmp/nvim/init.lua
006.893  000.354: sourcing vimrc file(s)
007.046  000.014  000.014: sourcing /usr/share/vim/vimfiles/ftdetect/lf.vim
007.093  000.167  000.153: sourcing /usr/local/share/nvim/runtime/filetype.lua
007.114  000.009  000.009: sourcing /usr/local/share/nvim/runtime/filetype.vim
007.221  000.052  000.052: sourcing /usr/local/share/nvim/runtime/syntax/synload.vim
007.303  000.164  000.112: sourcing /usr/local/share/nvim/runtime/syntax/syntax.vim
007.844  000.015  000.015: sourcing /usr/local/share/nvim/runtime/plugin/gzip.vim
007.870  000.010  000.010: sourcing /usr/local/share/nvim/runtime/plugin/health.vim
007.924  000.041  000.041: sourcing /usr/local/share/nvim/runtime/plugin/man.vim
007.955  000.014  000.014: sourcing /usr/local/share/nvim/runtime/plugin/matchit.vim
008.127  000.157  000.157: sourcing /usr/local/share/nvim/runtime/plugin/matchparen.vim
008.166  000.020  000.020: sourcing /usr/local/share/nvim/runtime/plugin/netrwPlugin.vim
008.295  000.006  000.006: sourcing /tmp/nvim/rplugin.vim
008.300  000.119  000.113: sourcing /usr/local/share/nvim/runtime/plugin/rplugin.vim
008.349  000.042  000.042: sourcing /usr/local/share/nvim/runtime/plugin/shada.vim
008.366  000.006  000.006: sourcing /usr/local/share/nvim/runtime/plugin/spellfile.vim
008.382  000.009  000.009: sourcing /usr/local/share/nvim/runtime/plugin/tarPlugin.vim
008.394  000.005  000.005: sourcing /usr/local/share/nvim/runtime/plugin/tohtml.vim
008.411  000.010  000.010: sourcing /usr/local/share/nvim/runtime/plugin/tutor.vim
008.427  000.009  000.009: sourcing /usr/local/share/nvim/runtime/plugin/zipPlugin.vim
008.801  000.328  000.328: sourcing /usr/share/vim/vimfiles/plugin/fzf.vim
009.559  000.395  000.395: require('base46')
009.732  000.106  000.106: require('base46.themes.onedark')
009.842  000.280  000.175: require('base46.integrations.defaults')
009.952  000.109  000.109: require('base46.glassy')
010.241  000.104  000.104: require('base46.integrations.statusline')
011.021  000.344  000.344: require('plugins.configs.others')
011.148  000.113  000.113: require('packer.load')
011.789  000.163  000.163: require('nvchad_ui.config')
011.792  000.760  000.483: require('nvchad_ui')
011.845  000.046  000.046: require('nvchad_ui.tabufline.lazyload')
011.997  003.161  001.122: sourcing /tmp/nvim/plugin/packer_compiled.lua
012.157  000.978: loading rtp plugins
012.275  000.119: loading packages
012.291  000.016: loading after plugins
012.303  000.012: inits 3
013.249  000.947: reading ShaDa
013.362  000.112: opening buffers
013.380  000.019: BufEnter autocommands
013.382  000.002: editing files in windows
013.393  000.011: VimEnter autocommands
013.395  000.002: UIEnter autocommands
013.583  000.150  000.150: sourcing /usr/local/share/nvim/runtime/autoload/provider/clipboard.vim
013.588  000.043: before starting main loop
013.858  000.062  000.062: require('nvchad_ui.statusline')
014.064  000.071  000.071: require('nvchad_ui.icons')
014.068  000.207  000.136: require('nvchad_ui.statusline.modules')
014.277  000.420: first screen update
014.278  000.002: --- NVIM STARTED ---
```

**Comment by siduck** [comment]
hmp! so does impatient speed up modules etc we defer?

**Comment by nullchilly** [comment]
Yes, every module we load are cached as long as it's loaded after impatient

**Comment by nullchilly** [comment]
@siduck If you want to improve startuptime even further why are you loading packer at startup?

**Comment by siduck** [comment]
> @siduck If you want to improve startuptime even further why are you loading packer at startup?
> 
> This whole chunk can be removed :/
> 
> ```lua
> 005.560  000.106  000.106: require('packer.util')
> 005.584  000.632  000.527: require('packer')
> 006.289  000.203  000.203: require('packer.log')
> 006.293  000.301  000.098: require('packer.async')
> 006.507  000.073  000.073: require('packer.result')
> 006.510  000.216  000.143: require('packer.jobs')
> 006.518  000.714  000.197: require('packer.plugin_utils')
> 006.693  000.168  000.168: require('packer.snapshot')
> 006.834  002.421  000.638: require('plugins')
> ```

i lazyloaded packer at its cmds only but I still saw packer in the startutime log :/ , is that the right way to lazyload it?

**Comment by siduck** [comment]
> Yes, every module we load are cached as long as it's loaded after impatient
> 
> Also remove `impatient.profile()` it is only there for statistic and cost an extra 0.1ms in event loop (I know, horribly slow :)

sounds awesome, i'll test it first locally and then add it :D

**Comment by nullchilly** [comment]
```lua
use { "wbthomason/packer.nvim", opt = true }
```

**Comment by nullchilly** [comment]
Also, you need to disable packer commands
```lua
packer.init {
	disable_commands = true,
}
```

**Comment by siduck** [comment]
hmm but it should work at cmds too right?

like 

```
cmd = {
"PackerInstall",
"PackerSync"
..
}
```

**Comment by nullchilly** [comment]
@siduck No because we didn't `require "plugins"` at startup

**Comment by Akianonymus** [comment]
@nullchilly Lazy loading impatient is like missing the whole point.

And forget about him getting to merge impatient ( or maybe not ), his useless startuptime logs ( which aren't even relevant for most of the lua stuff ) are too precious.

**Comment by siduck** [comment]
@Akianonymus not at all bruh, we lazy load crucial plugins like treesitter, nvim-lspconfig, indent-blankline etc on file_on_open thingy which means : 

- if current buffer doesnt have file then dont load 
- If it does then defer and load it 

We even defer our mappings & whichkey  stuff as thats heavy 

So loading impatient after vimenter will speeden up loading those defferred stuff that load after impatient

**Comment by siduck** [comment]
> @nullchilly Lazy loading impatient is like missing the whole point.
> 
> And forget about him getting to merge impatient ( or maybe not ), his useless startuptime logs ( which aren't even relevant for most of the lua stuff ) are too precious.

my startuptime logs are not useless, 1ms on this machine is like 15-20 ms on my old pentium machine. Each MS matters if saving it doesnt cost breaking huge functionality

**Comment by siduck** [comment]
> @siduck No because we didn't `require "plugins"` at startup

startup?

**Comment by nullchilly** [comment]
@siduck 
https://github.com/NvChad/NvChad/blob/main/init.lua#L10

Remove this line and put `require "plugins"` inside each packercmds

**Comment by siduck** [comment]
> @siduck
> 
> https://github.com/NvChad/NvChad/blob/df3ae31e941a06fd4fe22f565598626460979002/init.lua#L10
> 
> Remove this line and put `require "plugins"` inside each packercmds
> 
> Edit: Yeah I guess it should work

oh dang

**Comment by tomasky** [comment]
i like nvchad ,  because it starts fast and saves me time. So i think startuptime is not insignificant

**Comment by siduck** [comment]
> i like nvchad , because it starts fast and saves my time. So i think startuptime is not insignificant

if we can still save more time then why not xD

**Comment by nullchilly** [comment]
@siduck Did you forget about my pr or something D:

**Comment by siduck** [comment]
> @siduck Did you forget about my pr or something D:

nooo, i have to merge aki's PR first and improve some functions in it . 

Then i'll have to test this locally to check whether it really speedens up the deferred modules so it'll take a day or two!

**Comment by nullchilly** [comment]
@siduck You don't need to merge impatient right away

My pr is mostly about lazyload packer.nvim

**Comment by siduck** [comment]
@nullchilly startuptime has been increased cuz of recent commits :((

**Comment by siduck** [comment]
whichkey shows in the startuptime now and it takes lots of startuptime!

**Comment by siduck** [comment]
ok testing some parts of this PR 

with Impatient
![image](https://user-images.githubusercontent.com/59060246/183276143-485da830-69f2-4807-8c3a-7e8537b6e745.png)

**Comment by siduck** [comment]
![image](https://user-images.githubusercontent.com/59060246/183276321-0d550637-1176-4b57-8b23-015737f3ad5a.png)

only this was beneficial, try it yourself 

`hyperfine 'nvim +q'`

First do it with your way then try my way

**Comment by nullchilly** [comment]
@siduck Yeah I forgot all of them are `defer_fn 0`

Good work with the lazy module :)

**Comment by Akianonymus** [comment]
### why --startuptime is not relavant if imapatient is used ?

Let's understand this logically with layman terms

Let's assume 

x = amount of time that shows with --startuptime
y = unknown, amount of time take for rest of stuff that is not captured by --startuptime ( lua stuff, whatever )

now after merging impatient

x = previous x + 10 ms

y = still unknown but less 

now the problem is, how much less is y, if the decrease of y is greater than 10 ms ( which obviously it is, given you don't lazy load it ), then it is profitable, otherwise not.

I think that's enough to demonstrate why merging impatient is a great idea if the benchmarks on the readme of impatient is not just a farce.

**Comment by Akianonymus** [comment]
### why you should not lazy load it ?

Now, if you are convinced how --startuptime is not relavant if impatient is merged, let's come to the second point

The whole point of impatient is to cache lua modules, so early it is loaded, the more benefit you will get.

So, it should be actually at the top of our init.lua file and adding to that even our packer_compiled.lua should be loaded after impatient.

**Comment by nullchilly** [comment]
@Akianonymus 
Modules that are already fast enough don't need impatient to be fast
`x + impatient(x + y) < p(x) + impatient(y)`

Here is the statistic to back up my claims

For reference, my startuptime is 11ms->15ms without impatient and 25->34ms with impatient (Incredibly random because of sys)

In the test below I made sure to load every module
![image](https://user-images.githubusercontent.com/56817415/183277732-024dd424-8f0b-4f1d-9185-5a5fa324ef77.png)

Now let take a look at standard caches

# require impatient at the top
![image](https://user-images.githubusercontent.com/56817415/183277709-49cd58c1-42d8-46ce-9754-a9e759fa1066.png)


# lazyloading impatient

Standard method
```lua
_G.__luacache_config = {
  chunks = {
    enable = false,
    path = vim.fn.stdpath('cache')..'/luacache_chunks',
  },
  modpaths = {
    enable = false,
    path = vim.fn.stdpath('cache')..'/luacache_modpaths',
  }
}
```
![image](https://user-images.githubusercontent.com/56817415/183277559-16e3418d-d686-408e-8072-23cea38ee6d3.png)

Cache method
![image](https://user-images.githubusercontent.com/56817415/183277588-abba863d-3fb0-4211-b5a9-2adf461c769e.png)

**Comment by siduck** [comment]
we need impatient basically for the plugins which are loaded only when file is open + that plugin is deferred so crucial plugins that cant have actual lazyloading with cmd, event etc like lspconfig, treesitter etc are loaded after impatient loads, which results in faster deferring ( check the hyperfine benchmark above )
